### DISCUSSION TITLE
How to debug and modify codes

### DISCUSSION THREAD

**Discussion by mlin01** [danny-avila/LibreChat#4644] [Help Wanted]
Hi, 
We got LibreChat setup and runs perfect.  What we want to do is modify codes so that it will return a set of models to the user.  Users are classified into different categories in our db.  We think the code that we need to modify is in ModelService.js (let us know if that is the right place).  The problem we have is that we set up the breakpoints in the file.  No breakpoints were hit.  We have used VS code to debug frontend or backend, no breakpoint hit.

Looking forward to hear from you.

**Comment by alvieirajr** [comment]
Hi @mlin01. i just try to put a breakpoint on fetchOpenAIModels method on ModelServices.js file. The flow was paused on breakpoint.

![image](https://github.com/user-attachments/assets/395c75fb-ddbf-42c3-af86-41df3e81236f)

Maybe you miss something during prepare you enviroment to debug. 

**How i did it:**

1. Put the breckpoints.
2. npm run frontend 
3. Ctrl+Shit+D, and "RUN AND DEBUG" I choosen  "Node.js..." and chossen "Run Script backend". 
4. F5 to start debugin.
5. After login the breakpoint was hitten.

**Comment by mlin01** [comment]
Hi @alvieirajr ,
Thanks for the instructions.  Before I try to follow your instructions, I just run "npm run backend".  It gave compilation errors:
error: KeyvMongo connection error: Client network socket disconnected before secure TLS connection was established
error: There was an uncaught error: Client network socket disconnected before secure TLS connection was established

I did a bran new clone, the error still happened.  Is there due to windows update?  Please advice.

**Reply by alvieirajr** [reply]
Where are you mongo instance ? If you set in your mongo instance to use TLS you need put ?tls=true on URL from mongo instance (i think).

If you are using a mongo instance on cloud try check the firewall and permissions

**Reply by mlin01** [reply]
This is my mongo url -> MONGO_URI=mongodb+srv://mmlin23823:xxxxxxx@cluster0.0fe1c.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0
After I add ?tls=true, it still have the same error
<<
2024-11-14 18:23:12 error: KeyvMongo connection error: Client network socket disconnected before secure TLS connection was established
2024-11-14 18:23:12 error: There was an uncaught error: Client network socket disconnected before secure TLS connection was established
>>

**Reply by alvieirajr** [reply]
- I would try to test connection to your mongo ATLAS using command line:
`mongo "mongodb+srv://cluster0.0fe1c.mongodb.net" --username mmlin23823 --password xxxxxxx`
- Check your nodejs version, mongo Atlas require TLS 1.2+ to work (node 0.10+);
- Check if you firewall suport outgoing conections to 27017 port;
- You are using mongo ATLAS on a cluster, so you need to enable TLS/SSL to this cluster (I think that is on Security section in the Atlas UI);
- Check the dns resolution using `nslookup cluster0.0fe1c.mongodb.net
`

Anyway, at first glance, this  looks like a infrastructure problem and not a librechat issue.

**Reply by mlin01** [reply]
I got the compilation errors fixed.  It was IP address changed by my ISP.
Now, I use this launh.json below to run, no break point was hitted.  Can you please share your launch.json?
Thank you very much.
============================================================================
{
   "version": "0.2.0",
   "configurations": [
     {
       "type": "node",
       "request": "launch",
       "name": "Launch Program",
       "skipFiles": ["<node_internals>/**"],
       "program":  "api/server/index.js",
       "outFiles": ["${workspaceFolder}/**/*.js"],
       "preLaunchTask": "npm: backend:dev"
     }
   ],
   "tasks": {
     "version": "2.0.0",
     "tasks": [
       {
         "label": "npm: backend:dev",
         "type": "npm",
         "script": "backend:dev",
         "group": "build",
         "problemMatcher": [],
         "detail": "Run npm script backend:dev"
       }
     ]
   }
 }

**Comment by alvieirajr** [comment]
On my librechat\.vscode folder:

{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Launch LibreChat (debug)",
      "skipFiles": ["<node_internals>/**"],
      "program": "${workspaceFolder}/api/server/index.js",
      "env": {
        "NODE_ENV": "production"
      },
      "console": "integratedTerminal"
    }
  ]
}

on my .vscode folder on my root directory (The librechat is a subdirectory
inside my workspace):

{
    "configurations": [
        {
            "type": "msedge",
            "name": "ws://
127.0.0.1:9229/b2c3b450-38b1-425a-9ae9-c0136ba00bc2",
            "request": "launch",
            "url": "ws://127.0.0.1:9229/b2c3b450-38b1-425a-9ae9-c0136ba00bc2
"
        }
    ]
}

Em sex., 15 de nov. de 2024 às 20:02, mlin01 ***@***.***>
escreveu:

**Reply by mlin01** [reply]
@alvieirajr I am back on this project.  Can you give me a launch.json when Librechat is in the main folder?
In your launch.json file, I don't see the command "npm run backend:dev" and "npm run frontend", how they were triggered?

Also, can you kindly point how to get the user's login Id in ModelService.js file?

Thank you.

**Comment by mlin01** [comment]
Hi @alvieirajr ,

Thank you very much to solve my debugging issues.

Now, I am facing challenging requirement.  That is, customize LibreChat such that the login page use background color #15284b, after login successful, change the background color to white.
